python3.pkgs.sphinxcontrib-openapi: unbreak, switch to cilium fork

flokli opened the pull request:
###### Description of changes

This fork uses m2r instead of mdinclude.
 
m2r is unmaintained and incompatible with recent versions of mistune (and old ones are insecure).
 
See
 
 - https://github.com/cilium/cilium/commit/b9862461568dd41d4dc8924711d4cc363907270b
 - https://github.com/cilium/openapi/commit/cd829a05caebd90b31e325d4c9c2714b459d135f
 
for details.

This PR also packages `picobox` and `sphinx_mdinclude`, which are new dependencies of `sphinxcontrib-openapi`.

Due to `picobox` being [incompatible with Python 3.10 currently](https://github.com/ikalnytskyi/picobox/issues/55), this only unbreaks `python39.pkgs.sphinxcontrib-openapi`, but it's better than not being able to render openapi stuff at all.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

flokli committed:
python.pkgs.sphinx-mdinclude: init at 0.5.2

flokli committed:
python.pkgs.picobox: init

This adds a dependency needed for recent sphinxcontrib-openapi.

flokli committed:
python.pkgs.sphinxcontrib-openapi: switch to cilium fork

This fork uses m2r instead of mdinclude.

m2r is unmaintained and incompatible with recent versions of mistune.

See

 - https://github.com/cilium/cilium/commit/b9862461568dd41d4dc8924711d4cc363907270b
 - https://github.com/cilium/openapi/commit/cd829a05caebd90b31e325d4c9c2714b459d135f

for details.

flokli commented:
Hmmh, this still seems to fail to execute, showing a

> AttributeError: module 'collections.abc' has no attribute 'defaultdict'

…that's also visible in the (currently disabled) tests. I need to take a look.

flokli commented:
I got it to work! `sphinx` now is able to render openapi specs again.

Unfortunately, I wasn't able to get the test suite to parse, but considering it's also disabled for all other sphinx-related python packages too, it's probably not too much of an issue.

flokli commented:
This apparently got introduced in https://github.com/cilium/openapi/commit/a1d4fca2e7c3ae3cca69593baade1ebc297a12f (so I commented there).

We could probably just pin the commit before that instead of undoing the change in a patch but I'd rather see it fixed in `cilium/openapi `too.

flokli commented:
I sent a PR at https://github.com/cilium/openapi/pull/1.

SuperSandro2000 commented:
The prefix for python is ``python310.pkgs`` or ``python310Packages`` right now. 3.9 is still being build but not the default anymore and python2 is completely unsupported.

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -10324,6 +10326,8 @@ in {
 
   sphinx-navtree = callPackage ../development/python-modules/sphinx-navtree { };
 
+  sphinx_mdinclude = callPackage ../development/python-modules/sphinx_mdinclude { };
```
```suggestion
  sphinx-mdinclude = callPackage ../development/python-modules/sphinx-mdinclude { };
```
New package names get normalized according to the docs.

SuperSandro2000 commented on pkgs/development/python-modules/sphinxcontrib-openapi/default.nix:
```diff
@@ -1,26 +1,55 @@
+  patches = [
+    (
+      fetchpatch {
+        url = "https://github.com/cilium/openapi/pull/1/commits/184a41c90129397c07b1f69bd1a51ea3b94422d3.patch";
```
Isn't that patch missing a try catch?

flokli replied:
I don't understand the question. What would I try/catch here?

SuperSandro2000 commented on pkgs/development/python-modules/sphinxcontrib-openapi/default.nix:
```diff
@@ -1,26 +1,55 @@
   };
 
+  patches = [
+    (
+      fetchpatch {
+        url = "https://github.com/cilium/openapi/pull/1/commits/184a41c90129397c07b1f69bd1a51ea3b94422d3.patch";
+        hash = "sha256-SRhj6D6TKX6gW4mLMW+Bx85YjDFaKElDd9HKlPmfqq8=";
+      }
+    )
```
```suggestion
    (fetchpatch {
      url = "https://github.com/cilium/openapi/pull/1/commits/184a41c90129397c07b1f69bd1a51ea3b94422d3.patch";
      hash = "sha256-SRhj6D6TKX6gW4mLMW+Bx85YjDFaKElDd9HKlPmfqq8=";
    })
```

SuperSandro2000 commented on pkgs/development/python-modules/sphinx_mdinclude/default.nix:
```diff
@@ -0,0 +1,44 @@
+}:
+
+buildPythonPackage rec {
+  pname = "sphinx_mdinclude";
+  version = "0.5.1";
+  format = "flit";
+
+  src = fetchPypi {
+    inherit pname version;
```
```suggestion
  pname = "sphinx-mdinclude";
  version = "0.5.1";
  format = "flit";

  src = fetchPypi {
    pname = "sphinx_mdinclude";
    inherit version;
```

SuperSandro2000 commented on pkgs/development/python-modules/picobox/default.nix:
```diff
@@ -0,0 +1,47 @@
+  nativeBuildInputs = [
+    setuptools-scm
+  ];
+
+
```
```suggestion

```

SuperSandro2000 commented on pkgs/development/python-modules/picobox/default.nix:
```diff
@@ -0,0 +1,47 @@
+  pname = "picobox";
+  version = "2.2.0";
+
+  # currently broken on 3.10 due to https://github.com/ikalnytskyi/picobox/issues/55
+  disabled = isPy27 || pythonAtLeast "3.10";
```
```suggestion
  disabled = pythonOlder "3.7";
```
We can safely ignore those tests, the output did just change.

flokli replied:
I initially packaged this with tests, and then saw the some tests are broken, on Python 3.10, while they succeed on 3.9.

I'm not sure if openapi or other potential downstream consumers of this package make use of the (currently broken) functionality that picobox exposes, and rather mark it as broken for 3.10.

flokli replied:
As written further down, tests for python 3.10 are fixed by `fetchpatch`'ing another commit (not yet released). This is now not marked as broken for 3.10 anymore.

flokli commented:
> The prefix for python is `python310.pkgs` or `python310Packages` right now. 3.9 is still being build but not the default anymore and python2 is completely unsupported.

Yes, I used the python39 prefix in the PR to convey this unbreaks at least the python 3.9 version of the package - meaning if you assemble a python 3.9 environment with sphinx and this PR, you can render docs. Of course, once https://github.com/ikalnytskyi/picobox/issues/55 is fixed, we should relax the version bounds in picobox.

flokli commented:
Apparently, the python 3.10 fix for picobox is already in master, there just hasn't been a release yet. I `fetchpatch`'ed the patch and removed the python 3.9 restriction.

Likewise, `sphinx-mdinclude` [docutils support was fixed too](https://github.com/amyreese/sphinx-mdinclude/issues/8), so no need to ship a downstream patch in nixpkgs, we can fetchpatch this commit too.

In both cases, I linked to the issue/comment asking for a release containing the fixes.

I'd like to wait until https://github.com/cilium/openapi/pull/1 is merged before merging this PR in, so we can point to the final commit there, too.

flokli commented:
https://github.com/cilium/openapi/pull/1 has been merged, this PR was updated. This is ready from my side.

flokli commented:
sphinx-mdinclude [did just release 0.5.2, containing the patch](https://github.com/amyreese/sphinx-mdinclude/pull/9#issuecomment-1229139229), so I updated the version to this and dropped the patch.

SuperSandro2000 commented on pkgs/development/python-modules/picobox/default.nix:
```diff
@@ -0,0 +1,59 @@
+  pname = "picobox";
+  version = "2.2.0";
+
+  disabled = isPy27;
```
This should be pythonOlder "3.7"

SuperSandro2000 commented on pkgs/development/python-modules/sphinxcontrib-openapi/default.nix:
```diff
@@ -29,5 +49,4 @@ buildPythonPackage rec {
     description = "OpenAPI (fka Swagger) spec renderer for Sphinx";
     license = licenses.bsd0;
```
Missing maintainer

flokli replied:
https://github.com/NixOS/nixpkgs/pull/189112

SuperSandro2000 commented on pkgs/development/python-modules/sphinxcontrib-openapi/default.nix:
```diff
@@ -1,26 +1,46 @@
   pname = "sphinxcontrib-openapi";
-  version = "0.7.0";
+  version = "unstable-2022-08-26";
   disabled = isPy27;
```
This shoudl be pythonOlder "3.6"

flokli replied:
python 3.6 doesn't exist either. It works for all remaining non-python 2.7 versions in nixpkgs. Which should be fine.

qmonnet commented:
[Follow-up] @flokli Hi, the PR to upstream the changes from the cilium fork has been merged a few weeks ago (https://github.com/sphinx-contrib/openapi/pull/127), and a new tag has been created for sphinx/openapi (https://github.com/sphinx-contrib/openapi/releases/tag/0.8.0), so you probably want to revert to that source. We probably won't keep the cilium/ fork forever.

flokli commented:
@qmonnet thanks for the highlight. I opened https://github.com/NixOS/nixpkgs/pull/211156, which switches it back to the fork.
